### PR TITLE
Machine Learning Driven Filling

### LINKED ISSUE #139: Using Machine Learning For Data Filling
I'm building a class that will use machine learning (I'm looking at a variety of algorithms), to go through and fill in missing data. This is the approximate process,

Steps:
1. Create a new variable that is the merged version of the training and testing data
2. Separate all of the rows with missing data in them and place them in a list.
3. In the missing data list, a complete column is a predictor, an incomplete column is a target. Select a specific target for stages 4/5 and remove other target columns.
4. Train an algorithm using the complete data using the set of predictor columns in the missing data (train), and apply it to the incomplete data (test).
5. Repeat starting at step 3 if columns with missing data still exist.

I'm still tinkering with this and trying a few ways of doing it. Testing and documenting is going to take a while.

### PR DESCRIPTION
**Pull request by ASAAR**
closes #139 

Includes,
- A XGBoost based regression filler for missing data.
- Inclusion of XGBoost in the framework
- Replacement of launcher.py with simple call to hub.py

Caveat: I have temporarily removed the FillerStrategy from the worker. I'll be refactoring that some, with the goal of consolidating the two filling methods.

### CODE REVIEW

File: launcher.py [hunk 1]
```diff
@@ -1,57 +1,3 @@
...
-    if prediction_type != "regression" and prediction_type != "classification":
-        print("Please input a type of prediction to be doing")
-        return
-
-    run_framework(validation_pack, package, prediction_type)
-
-pre_ml_setup()
+from modules.hub import hub
 
+hub = hub.Hub()
```
Comment (asclines) [inline launcher.py]: Is it safe to have the variable name same as the object being imported?

File: modules/DMZ/data_kit/filler_regression.py [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
 
 class RegressionFiller(object):
 
     def __init__(self, train, test):
-        self.complete_data = self.select_complete_data(train)
-        self.incomplete_data = self.select_incomplete_data(train)
-        self.testing_data = test
-        self.select_missing_column(self.incomplete_data)
-        self.select_missing_column(self.complete_data)
+        self.model = None
```
Comment (asclines) [inline modules/DMZ/data_kit/filler_regression.py]: Whats the reasoning behind having these set to `None`?

File: modules/DMZ/data_kit/filler_regression.py [hunk 3]
```diff
@@ -1,15 +1,24 @@
...
+        self.filled_data = None
+        self.filled_test_data = None
+        self.train_data = train
+        self.test_data = test
+
+        self.total_data = pd.concat([train, test], ignore_index=True)
+        self.total_missing = self.select_incomplete_data(self.total_data)
+        self.total_data = self.select_complete_data(self.total_data)
+        self.filled_data = self.fill_missing_columns(self.train_data, self.total_data)
```
Comment (asclines) [inline modules/DMZ/data_kit/filler_regression.py]: So the `fill_missing_columns` gets called twice? Is it consistent? Or will there be a chance that the method would fill in the missing data each time its run?

File: modules/DMZ/data_kit/filler_regression.py [hunk 4]
```diff
@@ -19,14 +28,45 @@ def select_incomplete_data(self, data):
...
+            results = self.apply_filler(x_train, y_train, x_test)
+            data = self.fill_missing(y_test, results, data, target_col)
 
-    def train_filler(self):
-        return
+        return data
 
-    def fill_missing(self):
-        return
+    def get_filled_data(self):
```
Comment (asclines) [inline modules/DMZ/data_kit/filler_regression.py]: Why is this here? This seems very anti-Python-ey

File: modules/DMZ/models/classification/scikit_classification_learners.py [hunk 5]
```diff
@@ -20,3 +21,7 @@ def train_knn():
 def train_svc():
     # Selecting the model
     return mp.ModelProperties(), svm.SVC(decision_function_shape='ovo')
+
+
+def train_xgboost_classifier():
```
Comment (asclines) [inline modules/DMZ/models/classification/scikit_classification_learners.py]: Needs docs, at least mention that a tuple is being returned etc..

File: modules/DMZ/data_kit/filler_regression.py (hunk 2 above)
Comment (ZakeryFyke) [inline modules/DMZ/data_kit/filler_regression.py]: @ASAAR can correct me if I'm wrong, but AFAIK they need to be initialized.

File: modules/DMZ/data_kit/filler_regression.py (hunk 4 above)
Comment (RyanMcBerg) [inline modules/DMZ/data_kit/filler_regression.py]: Well object oriented python in general is kind of anti-Python-ey?

File: modules/DMZ/data_kit/filler_regression.py (hunk 2 above)
Comment (ASAAR) [inline modules/DMZ/data_kit/filler_regression.py]: Uh, yeah, that's accurate

File: modules/DMZ/data_kit/filler_regression.py (hunk 3 above)
Comment (ASAAR) [inline modules/DMZ/data_kit/filler_regression.py]: It's filling different sets of data, how is this an issue?

File: modules/DMZ/data_kit/filler_regression.py (hunk 4 above)
Comment (ASAAR) [inline modules/DMZ/data_kit/filler_regression.py]: Because it avoids problems if we refactor the class and change variable names

Comment (asclines) [inline modules/DMZ/data_kit/filler_regression.py]: if we change variable names the method names dont make too much sense either.

File: modules/DMZ/data_kit/filler_regression.py (hunk 3 above)
Comment (ASAAR) [inline modules/DMZ/data_kit/filler_regression.py]: Nope, because I split out the complete data and although I make predictions on the complete data, I never replace previously complete data

Comment (asclines) [inline modules/DMZ/data_kit/filler_regression.py]: Im probably just misunderstanding the behind the scenes on this.
